lib/rubocop/cops_documentation_generator.rb

```diff
@@ -0,0 +1,290 @@
+# frozen_string_literal: true
+
+# Class for generating documentation of all cops departments
+class CopsDocumentationGenerator # rubocop:disable Metrics/ClassLength
+  # This class will only generate documentation for cops that belong to one of
+  # the departments given in the `departments` array. E.g. if we only wanted
+  # documentation for Lint cops:
+  #
+  #   CopsDocumentationGenerator.new(departments: ['Lint']).call
+  #
+  def initialize(departments: [])
+    @departments = departments.map(&:to_sym).sort!
+    all_cops = RuboCop::Cop::Cop.registry
+    qualified_cops = all_cops.select { |cop| qualified_cop?(cop) }
+    @cops = RuboCop::Cop::Registry.new(qualified_cops)
+    @config = RuboCop::ConfigLoader.default_configuration
+  end
+
+  def call
+    YARD::Registry.load!
+    departments.each do |department|
+      print_cops_of_department(department)
+    end
+
+    print_table_of_contents
+  ensure
+    RuboCop::ConfigLoader.default_configuration = nil
+  end
+
+  private
+
+  attr_reader :departments, :cops, :config
+
+  def qualified_cop?(cop)
+    departments.include?(cop.department)
+  end
+
+  def cops_of_department(department)
+    cops.with_department(department).sort!
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  def cops_body(cop, description, examples_objects, pars)
+    content = h2(cop.cop_name)
+    content << required_ruby_version(cop)
+    content << properties(cop)
+    content << "#{description}\n"
+    content << examples(examples_objects) if examples_objects.count.positive?
+    content << configurations(pars)
+    content << references(cop)
+    content
+  end
+  # rubocop:enable Metrics/AbcSize
+
+  def examples(examples_object)
+    examples_object.each_with_object(h3('Examples').dup) do |example, content|
+      content << "\n" unless content.end_with?("\n\n")
+      content << h4(example.name) unless example.name == ''
+      content << code_example(example)
+    end
+  end
+
+  def required_ruby_version(cop)
+    return '' unless cop.respond_to?(:required_minimum_ruby_version)
+
+    "NOTE: Required Ruby version: #{cop.required_minimum_ruby_version}\n\n"
+  end
+
+  # rubocop:disable Metrics/MethodLength
+  def properties(cop)
+    header = [
+      'Enabled by default', 'Safe', 'Supports autocorrection', 'VersionAdded',
+      'VersionChanged'
+    ]
+    autocorrect = if cop.support_autocorrect?
+                    "Yes#{' (Unsafe)' unless cop.new(config).safe_autocorrect?}"
```

Comment:
`cop.new(config).safe_autocorrect?}`
Can this be fetched from the config instead? `cop_config.fetch('SafeAutoCorrect', true)`?

Reply:
It can: https://github.com/rubocop-hq/rubocop/blob/343f62e4555be0470326f47af219689e21c61a37/lib/rubocop/cop/autocorrect_logic.rb#L23-L26

Reply:
Now I see, makes sense 👍
The need to instantiate the cop triggered me, but since there's no other way around I guess it's totally fine.

Reply:
I mean, I think we could just call `cop_config.fetch('Safe', true) && cop_config.fetch('SafeAutoCorrect', true)`, but we’d be duplicating the logic and producing wrong results if the implementation of `#safe_autocorrect?` changes in the future. Would it be possible to move the logic into the `cop_config` object?

Reply:
> Would it be possible to move the logic into the `cop_config` object?

I'd second that.

Reply:
`cop_config` is a Hash, but it _might_ be possible to add the method by changing `RuboCop::Config#for_cop` to e.g.
```rb
@for_cop[cop.respond_to?(:cop_name) ? cop.cop_name : cop].tap do |config|
  def config.safe_autocorrect?
    fetch('Safe', true) && fetch('SafeAutoCorrect', true)
  end
end
```

Let’s consider this in a separate PR please. I think this PR is large enough as is.

Reply:
Oh, my bad, forget it then...

Reply:
😱